lib/active_publisher/async/in_memory_adapter/consumer_thread.rb

```diff
@@ -76,6 +84,35 @@ def start_thread
             end
           end
         end
+
+        def publish_all(channel, exchange_name, messages)
+          exchange = channel.topic(exchange_name)
+          potentially_retry = []
+          loop do
+            break if messages.empty?
+            message = messages.shift
+
+            fail ActivePublisher::UnknownMessageClassError, "bulk publish messages must be ActivePublisher::Message" unless message.is_a?(ActivePublisher::Message)
+            fail ActivePublisher::ExchangeMismatchError, "bulk publish messages must match publish_all exchange_name" if message.exchange_name != exchange_name
+
+            begin
+              options = ::ActivePublisher.publishing_options(message.route, message.options || {})
+              exchange.publish(message.payload, options)
+              potentially_retry << message
+            rescue
+              messages << message
+              raise
+            end
+          end
+          if channel.uses_publisher_confirms?
```

Comment:
This method has gotten pretty gnarly, but it feels weird to refactor it because we end up having multiple methods that are all modifying this `messages` argument. Not sure if that is better or worse than having one big message method at the edge of the code.